cats/cats-sql/src/main/kotlin/com/netflix/spinnaker/cats/sql/cache/SqlCacheCleanupAgent.kt

```diff
@@ -0,0 +1,131 @@
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.netflix.spinnaker.cats.sql.cache
+
+import com.netflix.spectator.api.Registry
+import com.netflix.spinnaker.cats.agent.RunnableAgent
+import com.netflix.spinnaker.cats.provider.ProviderRegistry
+import com.netflix.spinnaker.clouddriver.cache.CustomScheduledAgent
+import com.netflix.spinnaker.clouddriver.core.provider.CoreProvider
+import com.netflix.spinnaker.config.ConnectionPools
+import com.netflix.spinnaker.kork.sql.routing.withPool
+import org.jooq.DSLContext
+import org.jooq.impl.DSL.field
+import org.jooq.impl.DSL.table
+import org.slf4j.LoggerFactory
+import java.util.concurrent.TimeUnit
+
+/**
+ * Intermittently scans the entire database looking for records created by
+ * caching agents that are no longer configured.
+ */
+class SqlCacheCleanupAgent(
+  private val providerRegistry: ProviderRegistry,
+  private val jooq: DSLContext,
+  private val registry: Registry,
+  private val sqlNames: SqlNames
+) : RunnableAgent, CustomScheduledAgent {
+
+  private val log by lazy { LoggerFactory.getLogger(javaClass) }
+
+  private val deletedId = registry.createId("sql.cacheCleanupAgent.deleted")
+  private val timingId = registry.createId("sql.cacheCleanupAgent.timing")
+
+  override fun run() {
+    log.info("Scanning for cache records to cleanup")
+
+    val touchedResourceTables = mutableListOf<String>()
+    val touchedRelTables = mutableListOf<String>()
+
+    val agentTypes = providerRegistry.providers.flatMap { it.agents }.map { it.agentType }
+    withPool(ConnectionPools.CACHE_WRITER.value) {
+      agentTypes.forEach { agentType ->
+        val resourceTable = sqlNames.resourceTableName(agentType)
+        val relTable = sqlNames.relTableName(agentType)
+
+        // If [relTable] has not been touched yet, scan through each record it contains, deleting all records that do
+        // not correlate to a currently configured agent
+        if (!touchedRelTables.contains(relTable)) {
+          val rs = jooq.select(field("id"), field("agent"))
+            .from(table(relTable))
+            .fetch()
+            .intoResultSet()
+
+          val idsToClean = mutableListOf<String>()
+          while (rs.next()) {
+            if (!agentTypes.contains(rs.getString(2))) {
+              idsToClean.add(rs.getString(1))
+            }
+          }
+
+          if (idsToClean.isNotEmpty()) {
+            log.info("Found {} records to cleanup from '{}'", idsToClean.size, relTable)
+            idsToClean.chunked(100) { chunk ->
+              // TODO(rz): Add retry
```

Comment:
Looks like resilience4j is already used here in `sql-cats`, I would recommend doing it like this:

https://github.com/spinnaker/clouddriver/blob/master/clouddriver-sql/src/main/kotlin/com/netflix/spinnaker/clouddriver/sql/sql.kt#L36